### PR TITLE
Add wildreceipt dataset

### PR DESCRIPTION
**Pull request by HamzaGbada**

### CODE REVIEW

File: doctr/datasets/wildreceipt.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                        [
+                            [coordinates[0], coordinates[1]],
+                            [coordinates[2], coordinates[3]],
+                            [coordinates[4], coordinates[5]],
+                            [coordinates[6], coordinates[7]],
+                        ],
+                        dtype=np_dtype
+                    )
+                else:
+                    box = self._convert_xmin_ymin(coordinates)
```
Comment (felixdittrich92) [inline doctr/datasets/wildreceipt.py:101]: No need to write an own function you can use the functions from `doctr.utils`

```
from .utils import polygon_to_bbox
box_targets = polygon_to_bbox(tuple((coordniates[i], coordinates[i + 1]) for i in range(0, len(coordinates), 2)))
box = [coord for coords in box_targets for coord in coords]
```
OR

write the logic directly here (function is only used onces) 
```
x, y = box[::2], box[1::2]
box = [min(x), min(y), max(x), max(y)]
```

I would prefer the sec way
:)

File: doctr/datasets/wildreceipt.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                _targets.append((annotation['text'], box))
+            text_targets, box_targets = zip(*_targets)
+
+            if recognition_task:
+                crops = crop_bboxes_from_image(
+                    img_path=os.path.join(tmp_root, img_path), geoms=np.asarray(box_targets, dtype=int).clip(min=0)
+                )
+                for crop, label in zip(crops, list(text_targets)):
+                    with open(os.path.join(reco_folder_path, f"{reco_images_counter}.txt"), "w") as f:
```
Comment (felixdittrich92) [inline doctr/datasets/wildreceipt.py:110]: As mentioned i don't think that we need to save it locally wdyt ?

File: doctr/datasets/wildreceipt.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                        tmp_img = Image.fromarray(crop)
+                        tmp_img.save(os.path.join(reco_folder_path, f"{reco_images_counter}.png"))
+                        reco_images_counter += 1
+                    # self.data.append((crop, label))
+            else:
+                self.data.append(
+                    (img_path, dict(boxes=np.asarray(box_targets, dtype=int).clip(min=0), labels=list(text_targets)))
+                )
+        if recognition_task:
+            self._read_from_folder(reco_folder_path)
```
Comment (felixdittrich92) [inline doctr/datasets/wildreceipt.py:121]: same

File: doctr/datasets/wildreceipt.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        self.root = tmp_root
+
+    def extra_repr(self) -> str:
+        return f"train={self.train}"
+
+    def _read_from_folder(self, path: str) -> None:
```
Comment (felixdittrich92) [inline doctr/datasets/wildreceipt.py:127]: same

File: doctr/datasets/wildreceipt.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        for img_path in glob.glob(os.path.join(path, "*.png")):
+            with open(os.path.join(path, f"{os.path.basename(img_path)[:-4]}.txt"), "r") as f:
+                self.data.append((img_path, f.read()))
+
+    @staticmethod
+    def _convert_xmin_ymin(box: List) -> List:
```
Comment (felixdittrich92) [inline doctr/datasets/wildreceipt.py:133]: can be removed

**Review by felixdittrich92 — CHANGES_REQUESTED** [review]
Hi @HamzaGbada :wave:,

Thanks a lot this looks overall pretty good :+1: 
I have added a few comments

Furthermore could you also update the docs please ? :)
https://github.com/mindee/doctr/blob/main/docs/source/index.rst  -> Supported Datasets
https://github.com/mindee/doctr/blob/main/docs/source/using_doctr/using_datasets.rst -> Tables

If you are done please run
```
make style
make quality
```
to fix formatting, etc.

NOTE: Don't take care of the failing CI TF detection test i have opened a fix for this already :)

File: doctr/datasets/wildreceipt.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        if not os.path.exists(label_path) or not os.path.exists(img_folder):
+            raise FileNotFoundError(f"unable to locate {label_path if not os.path.exists(label_path) else img_folder}")
+
+        tmp_root = img_folder
+        self.train = train
+        np_dtype = np.float32
+        self.data: List[Tuple[Union[str, Path, np.ndarray], Union[str, Dict[str, Any]]]] = []
+
+        # define folder to write IMGUR5K recognition dataset
+        reco_folder_name = "WILDRECEIPT_recognition_train" if self.train else "WILDRECEIPT_recognition_test"
```
[the remark this answers is not in the document]
Comment (HamzaGbada) [reply doctr/datasets/wildreceipt.py:66]: Certainly, given the limited number of samples available – specifically, 1268 samples for the training set and 472 samples for the test set – I've opted to store the data directly in RAM.

**Comment by HamzaGbada** [comment]
About fixing formatting, these two commends return a Error:
```
make style
make quality
```

```
Sphinx error:
Builder name style not registered or available through entry point
make: *** [Makefile:20: style] Error 2
```
Do you have an idea about it ?

File: doctr/datasets/wildreceipt.py (hunk 1 above)
Comment (felixdittrich92) [reply doctr/datasets/wildreceipt.py]: @HamzaGbada can we use the secound suggestion please after reading this again i really don't like it 😅

Comment (HamzaGbada) [reply doctr/datasets/wildreceipt.py]: OK

**Comment by felixT2K** [comment]
> About fixing formatting, these two commends return a Error:
> 
> ```
> make style
> make quality
> ```
> 
> ```
> Sphinx error:
> Builder name style not registered or available through entry point
> make: *** [Makefile:20: style] Error 2
> ```
> 
> Do you have an idea about it ?

You have installed doctr with it's dev dependencies correct ?
```
cd doctr
pip3 install -e .[dev]
```

Looks like you are in the `docs` directory

```
cd doctr
make style
make quality
```
https://github.com/mindee/doctr/blob/main/Makefile

File: docs/source/using_doctr/using_datasets.rst [hunk 7]
```diff
@@ -84,6 +86,8 @@ This datasets contains the information to train or validate a text recognition m
 +-----------------------------+---------------------------------+---------------------------------+---------------------------------------------+
 | IIITHWS                     | 7141797                         | 793533                          | english / handwritten / external resources  |
 +-----------------------------+---------------------------------+---------------------------------+---------------------------------------------+
+| WILDRECEIPT                 | 1268                            | 472                             | english / external resources                |
```
Comment (felixT2K) [inline docs/source/using_doctr/using_datasets.rst:89]: This looks not correct here we should add the number of samples we get if we use the dataset for recognition :)
So this should be much more

File: doctr/datasets/wildreceipt.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+from ..utils import polygon_to_bbox
+
+
+class WILDRECEIPT(AbstractDataset):
+    """WildReceipt dataset from `"Spatial Dual-Modality Graph Reasoning for Key Information Extraction"
+        <https://arxiv.org/abs/2103.14470v1>`_ |
+    `repository <https://download.openmmlab.com/mmocr/data/wildreceipt.tar>`_.
+
+    >>> # NOTE: You need to download the dataset from the repository.
```
Comment (felixT2K) [inline doctr/datasets/wildreceipt.py:28]: You need to download the dataset first.

File: doctr/datasets/wildreceipt.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+                    x, y = coordinates[::2], coordinates[1::2]
+                    box = [min(x), min(y), max(x), max(y)]
+                _targets.append((annotation['text'], box))
+            text_targets, box_targets = zip(*_targets)
+
+            if recognition_task:
+                crops = crop_bboxes_from_image(
+                    img_path=os.path.join(tmp_root, img_path), geoms=np.asarray(box_targets, dtype=int).clip(min=0)
+                )
+                for crop, label in zip(crops, list(text_targets)):
```
Comment (felixT2K) [inline doctr/datasets/wildreceipt.py:101]: Do you know if there are text inside we need to filter out ? 
For example text which contains whitespaces ?

Ref.: https://github.com/mindee/doctr/blob/f22f6dd4e038082a90bb6a7b78ffe8a356368fc7/doctr/datasets/funsd.py#L100

File: doctr/datasets/wildreceipt.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+__all__ = ["WILDRECEIPT"]
+
+from ..utils import polygon_to_bbox
+
+
+class WILDRECEIPT(AbstractDataset):
+    """WildReceipt dataset from `"Spatial Dual-Modality Graph Reasoning for Key Information Extraction"
+        <https://arxiv.org/abs/2103.14470v1>`_ |
+    `repository <https://download.openmmlab.com/mmocr/data/wildreceipt.tar>`_.
+
```
Comment (felixT2K) [inline doctr/datasets/wildreceipt.py:27]: **Optional**
If we have an image to give a general overview of the dataset would be great

See: 
https://mindee.github.io/doctr/modules/datasets.html
https://github.com/mindee/doctr/blob/f22f6dd4e038082a90bb6a7b78ffe8a356368fc7/doctr/datasets/funsd.py#L24

**Review by felixT2K — CHANGES_REQUESTED** [review]
@HamzaGbada Close to merge really good job 👍🏼 😃

File: doctr/datasets/wildreceipt.py (hunk 9 above)
Comment (HamzaGbada) [reply doctr/datasets/wildreceipt.py:101]: t's worth noting that this dataset contains small text elements that might not be conducive to the recognition task. For instance, we could consider filtering out text elements that are empty or consist of characters such as `"-", "*", "/", "=", "#", or "@"` to enhance the quality of the recognition process.

File: doctr/datasets/wildreceipt.py (hunk 10 above)
Comment (HamzaGbada) [reply doctr/datasets/wildreceipt.py:27]: where should I put the image ?

File: doctr/datasets/wildreceipt.py (hunk 9 above)
Comment (felixT2K) [reply doctr/datasets/wildreceipt.py:101]: @HamzaGbada 
Mh in this case i think it would be enough to filter empty elements or if a whitespace is in the label.
We can handle all the above punctuations :)

File: doctr/datasets/wildreceipt.py (hunk 10 above)
Comment (felixT2K) [reply doctr/datasets/wildreceipt.py:27]: @HamzaGbada you can post it here

@odulcy-mindee  Could you upload it please ?

**Comment by felixT2K** [comment]
It would be enough if you post the mentioned image here we can update the docstring later :)
Does make style and quality now work ?

File: doctr/datasets/wildreceipt.py (hunk 10 above)
Comment (HamzaGbada) [reply doctr/datasets/wildreceipt.py:27]: combined_image

**Comment by HamzaGbada** [comment]
> It would be enough if you post the mentioned image here we can update the docstring later :) Does make style and quality now work ?

No it returns:

```
isort .
make: isort: No such file or directory
make: *** [Makefile:12: style] Error 127
```

**Comment by felixT2K** [comment]
> > It would be enough if you post the mentioned image here we can update the docstring later :) Does make style and quality now work ?
> 
> No it returns:
> 
> ```
> isort .
> make: isort: No such file or directory
> make: *** [Makefile:12: style] Error 127
> ```

what happens if you run the following commands (single without make):

```
isort .
black .
ruff --fix .
```

File: doctr/datasets/wildreceipt.py [hunk 11]
```diff
@@ -99,7 +99,7 @@ def __init__(
                     img_path=os.path.join(tmp_root, img_path), geoms=np.asarray(box_targets, dtype=int).clip(min=0)
                 )
                 for crop, label in zip(crops, list(text_targets)):
-                    if not any(char in label for char in ["", "-", "*", "/", "=", "#", "@"]):
+                    if not any(char in label for char in ["", " "]):
```
Comment (felixT2K) [inline doctr/datasets/wildreceipt.py:102]: `if label and " " not in label:`

**Comment by HamzaGbada** [comment]
> ruff --fix .

Got it, the issue was related to my Linux distribution.

**Comment by felixdittrich92** [comment]
@HamzaGbada 
black one change left see: 
https://github.com/mindee/doctr/actions/runs/6666409340/job/18117882074?pr=1359

**Review by felixdittrich92 — APPROVED** [review]
Looks good now thanks a lot :hugs:

File: doctr/datasets/wildreceipt.py (hunk 10 above)
Comment (odulcy-mindee) [reply doctr/datasets/wildreceipt.py:23]: @felixT2K @HamzaGbada Here you go:
```
https://doctr-static.mindee.com/models?id=v0.7.0/wildreceipt-dataset.jpg&src=0
```

**Comment by odulcy-mindee** [comment]
Thank you @HamzaGbada for this contribution ! :clap:
Thanks @felixdittrich92 for the review !
